src/data_generators/copula_data_set.py

```diff
@@ -0,0 +1,207 @@
+
+    def ppf(self, p: float) -> int:
+        """Calculates ppf of the count distribution defined by the histogram.
+
+        ppf is the inverse of cdf as will be explained in "Returns".
+
+        Args:
+            p:  A probability.
+
+        Returns:
+            A count x for which cdf[x] <= p and cdf[x + 1] > p, where cdf is
+            the cumulative sum of self.hist.
+        """
+        return np.searchsorted(a=self.cdf, v=p)
+
+
+class CopulaDataSet(DataSet):
+    """A DataSet constructed by gluing a list of PublisherData with Copula."""
+
+    def __init__(
+        self,
+        unlabeled_publisher_data_list: Iterable[PublisherData],
+        copula_generator: Copula = GaussianCopula(corr=0),
+        universe_size: int = None,
+        pricing_generator: PricingGenerator = FixedPriceGenerator(0.1),
+        random_generator: np.random.Generator = np.random.default_rng(0),
+        name: str = "copula",
+    ):
+        """Constructor for OverlapDataSet.
+
+        Args:
+            unlabeled_publisher_data_list:  A list of PublisherData. Each PublisherData
+                indicates the frequency distribution at a publisher --- the
+                ids labels are meaningless while their distribution matters.
+            copula_generator:  An instance of a subclass of
+                statsmodels.distributions.copula.copulas.Copula.  See here for the
+                possible choices:
+                https://github.com/statsmodels/statsmodels/tree/32dc52699f994acbf9dbdb9bd10d7eff04d860f5/statsmodels/distributions
+            universe_size:  A cross-publisher universe size to construct the copula.
+            pricing_generator:  A PricingGenerator object that annotates a list of
+                id's with randomly generated price information.  At this moment,
+                assume that all the publishers share the same PricingGenerator.
+            random_generator:  A random generator to draw sample from the copula.
+            name:  If specified, a human-readable name that will be associated to this
+                DataSet.
+        """
+        if universe_size is None:
+            universe_size = 2 * max(
+                [1] + [data.max_reach for data in unlabeled_publisher_data_list]
+            )
+        self.marginal_distributions = [
+            AnyFrequencyDistribution(self.zero_included_pmf(pub, universe_size))
+            for pub in unlabeled_publisher_data_list
+        ]
+        self.copula_generator = copula_generator
+        self.distribution = CopulaDistribution(
+            copula=self.copula_generator,
+            marginals=self.marginal_distributions,
+        )
+        self.sample = self.distribution.rvs(
+            nobs=universe_size,
+            random_state=random_generator.integers(0, 1e9),
+        ).astype("int32")
+        super().__init__(
+            publisher_data_list=[
+                PublisherData(
+                    impression_log_data=pricing_generator(pub_imps),
```

Comment:
Should you first shuffle pub_imps?

Reply:
You are right.  Thanks for catching this important bug!
Added shuffle in lines 129-131